File: May 2022/May_21.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        sort(coins.begin(), coins.end());
        int n = coins.size();
        vector<vector<int> > dp(n, vector<int>(amount + 1, - 1));
        int ans = helper(coins, amount, n - 1, dp);
        return ans >= 100 ? -1 : ans;
    }
    
    int helper(vector<int>& coins, int amount, int i, vector<vector<int> >& dp){
        if(amount == 0)
            return 0;
        if(amount < 0 || i < 0)
            return 100;
        if(dp[i][amount] != -1)
            return dp[i][amount];
        return dp[i][amount] = min(1 + helper(coins, amount - coins[i], i, dp), helper(coins, amount, i - 1, dp));
    }
};

File: May 2022/May_21.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        Arrays.sort(coins);
        dp = new Integer[amount + 1][coins.length];
        int ans = helper(coins, amount, coins.length - 1);
        return ans >= 100 ? -1 : ans;
    }
    
    Integer[][] dp;
    
    private int helper(int[] coins, int amount, int i){
        if(amount == 0)
            return 0;
        if(amount < 0 || i < 0)
            return 100;
        if(dp[amount][i] != null)
            return dp[amount][i];
        return dp[amount][i] = Math.min(1 + helper(coins, amount - coins[i], i), helper(coins, amount, i - 1));
    }
}
